### PR TITLE
Allow Update to use custom workflow lease

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

(1) Allow to run an Update operation with a custom `WorkflowLease`.
(2) Separate the WorkflowLease-specific step from the post-WorkflowLease steps.

## Why?

This is a refactoring to enable Update-With-Start. It requires these two abilities mentioned above.

## How did you test it?

Existing test.

### CODE REVIEW

File: service/history/api/updateworkflow/api.go [hunk 1]
```diff
@@ -61,23 +61,65 @@ const (
...
+	})
+	return commitFn()
+}
 
-	// Variables shared with wfCtxOperation. Using values instead of pointers to make sure
-	// they are copied and don't have any pointers to workflow context or mutable state.
+//nolint:revive
+func (u *Updater) Apply(
+	ctx context.Context,
+	withLease func(api.UpdateWorkflowActionFunc) error,
```
Comment (bergundy) [inline service/history/api/updateworkflow/api.go:117]: Name this `action`?

**Review by bergundy — APPROVED** [review]

File: service/history/api/updateworkflow/api.go [hunk 2]
```diff
@@ -183,102 +221,96 @@ func Invoke(
...
-				req,
-				&updatepb.Outcome{
-					Value: &updatepb.Outcome_Failure{Failure: update.CancelReasonWorkflowCompleted.RejectionFailure()},
-				},
-				enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED)
-			return rejectionResp, err
+	return func() (*historyservice.UpdateWorkflowExecutionResponse, error) {
+		// Wrapping workflow context related operation in separate func to prevent usage of its fields
+		// (including any mutable state fields) outside of this func after workflow lock is released.
+		// It is important to release workflow lock before calling matching.
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply service/history/api/updateworkflow/api.go:229]:
```suggestion
```

**Comment by stephanos** [comment]
Based on a conversation with Alex, I decided to move the variables into the struct to make the code easier to understand.

File: service/history/api/updateworkflow/api.go [hunk 3]
```diff
@@ -61,177 +61,204 @@ const (
...
-	req *historyservice.UpdateWorkflowExecutionRequest,
+type Updater struct {
+	shardCtx                   shard.Context
+	workflowConsistencyChecker api.WorkflowConsistencyChecker
+	matchingClient             matchingservice.MatchingServiceClient
+	req                        *historyservice.UpdateWorkflowExecutionRequest
+	namespaceID                namespace.ID
+
+	// Variables from workflow.Context and mutable state.
+	// NOTE: They *have to* be copies to avoid data races when using them outside the workflow lease.
```
Comment (alexshtin) [inline service/history/api/updateworkflow/api.go:72]: Not all of them. `wfKey`, `upd`, and `directive` can be used outside of lease. `directive` is very tricky though.

File: service/history/api/updateworkflow/api.go [hunk 4]
```diff
@@ -61,177 +61,204 @@ const (
...
+}
+
+func (u *Updater) OnError(
+	err error,
+) (*historyservice.UpdateWorkflowExecutionResponse, error) {
+	// If update is received while WFT is running, it will be waiting for the next WFT.
+	// And if that running WFT completes workflow, then update is rejected (see CancelIncomplete).
+	// Special handling for consts.ErrWorkflowCompleted here is needed to keep parity with this.
+	// I.e. if update is received and workflow was completed, or is about to be completed,
+	// then update is consistently rejected (instead of returning error in some cases).
```
Comment (alexshtin) [inline service/history/api/updateworkflow/api.go:225]: I need to rephrase this. I wrote it and had to read 2 times to get what it is about.

**Review by alexshtin — APPROVED** [review]

File: service/history/api/updateworkflow/api.go (hunk 3 above)
Comment (stephanos) [reply service/history/api/updateworkflow/api.go:72]: Better safe than sorry? :)

File: service/history/api/updateworkflow/api.go (hunk 4 above)
Comment (alexshtin) [reply service/history/api/updateworkflow/api.go:227]:
```suggestion
	// Special handling for consts.ErrWorkflowCompleted here is needed for consistency with the case when update is received while WFT is running and this WFT completes workflow. In this case update is rejected (see update.CancelIncomplete).
```

**Pull request merged**